pandas/tests/frame/indexing/test_setitem.py

```diff
@@ -379,6 +379,15 @@ def test_setitem_complete_column_with_array(self):
         assert expected["d"].dtype == arr.dtype
         tm.assert_frame_equal(df, expected)
 
+    def test_setitem_periodd_dtype(self):
+        # GH 39763
+        rng = period_range("2016-01-01", periods=9, freq="D", name="A")
+        df = DataFrame(rng)
+        expected = df.dtypes
+        df.iloc[:] = rng._na_value
+        result = df.dtypes
+        tm.assert_equal(result, expected)
```

Comment:
Please construct the expected DataFrame and use `tm.assert_frame_equal`

Reply:
> Please construct the expected DataFrame and use tm.assert_frame_equal

I'm a bit confused here, I thought that way but after executing this line `df.iloc[:] = rng._na_value` we get following o/p

```py
# o/p of  df.iloc[:] = rng._na_value
	A
0	NaT
1	NaT
2	NaT
3	NaT
4	NaT
5	NaT
6	NaT
7	NaT
8	NaT

# And it's data type
# -> df.dtypes
A    period[D]
dtype: object
```
and that was a  `result`.

but for `expected` when I tried creating the same data frame and changing its type to `period[D]` it did not work.
```py
expected = pd.DataFrame({"A": ['NaT', 'NaT', 'NaT', 'NaT', 'NaT', 'NaT', 'NaT', 'NaT', 'NaT']})
expected.astype(obj.dtypes) # 1 did not work
expected.astype("period[D]") # 2 did not work
expected.dtypes

# o/p
A    object
dtype: object
```

as per your suggestion, we can create a new data frame using period_range and compare using `tm.assert_frame_equal` here dtypes would be the same but the data is different so the test will fail.

I hope I'm able to describe the confusion clearly here.
Please let me know your thoughts on this.

Reply:
```
In [3]: pd.DataFrame({"A": ["NaT"]}, dtype="period[D]")
Out[3]: 
     A
0  NaT

In [4]: pd.DataFrame({"A": ["NaT"]}, dtype="period[D]").dtypes
Out[4]: 
A    period[D]
dtype: object
```